Tutorial interferometer (#151)
XanaduAI/strawberryfields: doc/tutorials/starship.rst

> 

Author: Nope, there are all just standard ReST files. Just a matter of different contributors using different naming conventions. Since it doesn't show in the docs, I'm fine with leaving it as is.

   You have successfully authenticated to the platform!

.. _first_program:

Executing your first program
============================

The easiest way to execute a program using StarshipEngine is to create a Blackbird script (an ``xbb`` file)
and place it in your current working directory. Check the :ref:`Blackbird tutorial <blackbird>` for how to create this file.

For this example, consider the following Blackbird script, which represents a quantum program that matches

> exactly the gate layout of the Chip0 photonic hardware device. We will save the following file as ``test.xbb``

Author: This is intentional :) At the moment, this is the only place where the chip is mentioned.

in our current working directory:

.. code-block:: python

   name template_2x2_chip0     # Name of the program

>    version 1.0                 # Blackbird version number

Author: Yep - this is the version of the Blackbird language, i.e., the overall grammar, syntax, and intrinsic functions etc. If there are any changes to Blackbird as a language, we will increment the version number here. This will allow the parser to support multiple versions of the Blackbird language.

   target chip0 (shots = 50)   # This program will run on chip0 for 50 shots

> 

Author: This gives the impression that the on-chip squeezing is a parametrized value, that anyone can modify; however, currently the squeezing value is fixed to 1.0. I moved it directly into the S2gate to avoid bugs due to people tempted to change the squeezing value :)

Of course, it won't fully stop anyone from changing the squeezing magnitude, just makes it a little less visible.

> 

Reviewer: Yep makes sense!

   # Define the interferometer phase values
   float phi0 = 0.574
   float phi1 = 1.33

>    # Initial states are two-mode squeezed states,

Author: ```suggestion
   # Initial states are two-mode squeezed states
```

>    # applied to alternating pairs of modes.

Author: Ah, I meant that the modes look like `[0, 2], [1, 3]`. Is there a better way I could phrase it? Might remove it altogether.
```suggestion
```

>    # applied to alternating pairs of modes.

Reviewer: Yeah I think it's better to leave out than add a potentially ambiguous sentence. I also can't think of a clearer way to say it succinctly!

   S2gate(1.0, 0.0) | [0, 2]
   S2gate(1.0, 0.0) | [1, 3]

   # A standard 2x2 interferometer is applied
   # to the first pair of modes

>    Rgate(phi0) | [0]

Author: Just the structure of the chip. The chip uses Mach-Zehnder interferometers, consisting of two parametrized phase-shifters, and two fixed 50-50 beamsplitters. Since the beamsplitters are not tunable, the transmissivity and phase are determined by the internal and external phase shifters.

   BSgate(pi/4, pi/2) | [0, 1]
   Rgate(phi1) | [0]
   BSgate(pi/4, pi/2) | [0, 1]

   # The 2x2 interferometer above is duplicateed
   # for the second pair of modes
   Rgate(phi0) | [2]
   BSgate(pi/4, pi/2) | [2, 3]
   Rgate(phi1) | [2]
   BSgate(pi/4, pi/2) | [2, 3]

>    # Perform a PNR measurement in the Fock basis

Reviewer: ```suggestion
   # Perform a photon number counting measurement
```
Just in case it is not clear what a PNR is?

   result = eng.run(prog)
   print(result.samples)

Executing your Blackbird script from the command line
-----------------------------------------------------

To execute this file from the command line, use the ``starship`` command as follows:

.. code-block:: console

   starship --input test.xbb --output out.txt

After executing the above command, the result will be stored in ``out.txt`` in the current working directory.
You can also omit the ``--output`` parameter to print the result to the screen.

Program compilation
===================

In addition to using the program template above, which directly matches the physical

> layout of the hardware device, you can define your interferometer using any combination

Reviewer: Maybe: "..., you can apply any two-mode interferometer to the pairs of modes. The interferometer can be composed of any combination ..."

Author: ```suggestion
layout of the hardware device, you can apply any two-mode interferometer to the pairs of modes.
The interferometer can be composed of any combination
```

of beamsplitters (:class:`~.ops.BSgate`), rotations/phase shifts (:class:`~.ops.Rgate`).

> Additionally, you can use the :class:`~.ops.Interferometer` command to directly pass a

Reviewer: ```suggestion
Furthermore, you can use the :class:`~.ops.Interferometer` command to directly pass a
```
Just to remove the repeat of "addition"

For example, consider the following Blackbird script:

.. code-block:: python

   name compilation_example  # Name of the program
   version 1.0               # Blackbird version number
   target chip0 (shots=50)   # This program will run on chip0 for 50 shots

   # Define a unitary matrix
   complex array U[2, 2] =
      -0.1955885-0.16833594j, 0.77074506+0.58254631j
      -0.03596574+0.96546083j, 0.00676031+0.2579654j

   # Initial states are two-mode squeezed states,
   # applied to alternating pairs of modes.
   S2gate(1.0, 0.0) | [0, 2]
   S2gate(1.0, 0.0) | [1, 3]

   # Apply the unitary matrix above to
   # the first pair of modes, as well
   # as a beamsplitter
   Interferometer(U) | [0, 1]
   BSgate(0.543, -0.123) | [0, 1]

   # Duplicate the above unitary for
   # the second pair of modes
   Interferometer(U) | [2, 3]
   BSgate(0.543, -0.123) | [2, 3]

   # Perform a PNR measurement in the Fock basis
   MeasureFock() | [0, 1, 2, 3]

.. note:: You may use :func:`~.random_interferometer` to generate arbitrary random unitaries.

This program will execute following the same steps as above; ``StarshipEngine`` will automatically

> compile the program to match the layout of the chip described in :ref:`first_program`.

Reviewer: Very cool!

You may wish to view the compiled program; this can be easily done in Python using
the :meth:`~.Program.compile` method:

>>> from strawberryfields import StarshipEngine
>>> from strawberryfields.io import load
>>> prog = load("test.xbb")
>>> prog = prog.compile("chip0")
>>> prog.print()
S2gate(1, 0) | (q[0], q[2])
S2gate(1, 0) | (q[1], q[3])
Rgate(0.9355) | (q[0])
BSgate(0.7854, 1.571) | (q[0], q[1])
Rgate(4.886) | (q[0])
BSgate(0.7854, 1.571) | (q[0], q[1])
Rgate(-0.3742) | (q[0])
Rgate(-0.05099) | (q[1])
Rgate(0.9355) | (q[2])
BSgate(0.7854, 1.571) | (q[2], q[3])
Rgate(4.886) | (q[2])
BSgate(0.7854, 1.571) | (q[2], q[3])
Rgate(-0.3742) | (q[2])
Rgate(-0.05099) | (q[3])

> MeasureFock | (q[0], q[1], q[2], q[3])

Reviewer: :thinking: It doesn't seem to match exactly the version above. There are, e.g., too many `Rgate`'s. Is this ok?

Author: Ah good point, the one above is outdated and should match this one.

Reviewer: Great, let me know if you need a hand with that bit. Otherwise looks ready to go for me!

Author: Just pushed a commit to update the first Blackbird codeblock
